pkg/httpd/websftp.go

```diff
@@ -0,0 +1,201 @@
+		h.handleAction(h.rename, request, &message)
+	case "mkdir":
+		h.handleAction(h.mkdir, request, &message)
+	default:
+		message.Err = "Unknown command"
+		h.ws.SendMessage(&message)
+	}
+}
+
+func (h *webSftp) handleList(request *webSftpRequest, response *Message) {
+	response.Data = h.list(request.Path)
+	response.CurrentPath = h.currentPath
+	h.ws.SendMessage(response)
+}
+
+func (h *webSftp) list(path string) string {
+	files := h.volume.List(path)
+	h.currentPath = h.volume.UserSftp.GetCurrentPath()
+	data, _ := json.Marshal(files)
+	return string(data)
+}
+
+func (h *webSftp) handleDownload(request *webSftpRequest, response *Message) {
+	file, filename, err := h.volume.Download(request.Path, request.IsDir)
+	if err != nil {
+		response.Err = err.Error()
+		h.ws.SendMessage(response)
+		return
+	}
+
+	if file.Reader != nil {
+		defer file.Reader.Close()
+	}
+
+	h.streamFileContent(file, response)
+	response.Data = filename
+	response.Type = SFTPData
+	h.ws.SendMessage(response)
+}
+
+func (h *webSftp) streamFileContent(file FileData, response *Message) {
+	response.Type = SFTPBinary
+	buf := make([]byte, 1024*64)
+	for {
+		select {
+		default:
+			responseCopy := *response
+			n, err := file.Reader.Read(buf)
+			if err != nil {
+				if err != io.EOF {
+					logger.Errorf("Error reading file: %s", err)
+					responseCopy.Err = err.Error()
+					h.ws.SendMessage(&responseCopy)
+				}
+				responseCopy.Raw = append([]byte{}, buf[:n]...)
+				h.ws.SendMessage(&responseCopy)
+				return
+			}
+
+			responseCopy.Raw = append([]byte{}, buf[:n]...)
+			h.ws.SendMessage(&responseCopy)
+		}
+	}
+}
+
+func (h *webSftp) handleUpload(request *webSftpRequest, msg *Message, response *Message) {
+	reader := bytes.NewReader(msg.Raw)
+	var readerAt io.ReaderAt = reader
+
+	id, idErr := strconv.Atoi(msg.Id)
+	if idErr != nil {
+		response.Err = idErr.Error()
+		h.ws.SendMessage(response)
+		return
+	}
+	var err error
+	if request.Merge {
+		err = h.volume.MergeChunk(id, request.Path)
+	} else if request.Chunk {
+		err = h.volume.UploadChunk(id, request.Path, request.OffSet, int64(reader.Len()), readerAt)
+	} else {
+		err = h.volume.UploadFile(request.Path, readerAt, request.Size)
+	}
+	if err != nil {
+		response.Err = err.Error()
+		h.ws.SendMessage(response)
+		return
+	}
+	response.Data = request.Path
+	h.ws.SendMessage(response)
+}
+
+func (h *webSftp) handleAction(action func(*webSftpRequest) error, request *webSftpRequest, response *Message) {
+	err := action(request)
+	if err != nil {
+		response.Err = err.Error()
+	} else {
+		response.Data = "ok"
+	}
+	h.ws.SendMessage(response)
+}
+
+func (h *webSftp) rm(request *webSftpRequest) error {
+	return h.volume.Remove(request.Path)
+}
+
+func (h *webSftp) rename(request *webSftpRequest) error {
+	oldNamePath := request.Path
+	newName := request.NewName
+	return h.volume.Rename(oldNamePath, newName)
+}
+
+func (h *webSftp) mkdir(request *webSftpRequest) error {
+	return h.volume.MakeDir(request.Path)
+}
```

Comment:
Here is an updated version of the code with minor improvements:

```go
@@ -0,0 +139,7 @@
    import (
        "bytes"
        "encoding/json"
        "fmt"
        "io/fs"
        "log"

        // Import other dependencies here...
```

To maintain consistency and readability throughout the codebase, we will remove unnecessary comments.

The existing files seem to have no syntax errors but might contain some logical flaws. You will need a developer with more expertise in Go programming to review it further.